discord.go

```diff
@@ -50,80 +50,97 @@ func discordInit() {
 		return
 	}
 
-	var webhook *discordgo.Webhook
+	var webhooks []*discordgo.Webhook
 	// get or create a webhook if we're not in compact mode
 	if !Integrations.Discord.CompactMode {
-		webhooks, err := sess.ChannelWebhooks(Integrations.Discord.ChannelID)
+		webhooks, err = sess.ChannelWebhooks(Integrations.Discord.ChannelID)
 		if err != nil {
 			Log.Println("Error getting Discord webhooks:", err)
-			return
-		}
-		for _, wh := range webhooks {
-			if wh.Name == "Devzat" {
-				webhook = wh
-			}
-		}
-		if webhook == nil {
-			webhook, err = sess.WebhookCreate(Integrations.Discord.ChannelID, "Devzat", "")
-			if err != nil {
-				Log.Println("Error creating a Discord webhook:", err)
-				return
-			}
+			webhooks = make([]*discordgo.Webhook, 0, maxWebhookCount)
 		}
 	}
 	DiscordChan = make(chan DiscordMsg, 100)
-	editsInLastMinute := 0 // discord allows for 30 webhook edits per minute: https://twitter.com/lolpython/status/967621046277820416
 	go func() {
-		overloading := false
+	nextMsg:
 		for msg := range DiscordChan {
-			sendingTimeStart := time.Now()
 			txt := strings.ReplaceAll(msg.msg, "@everyone", "@\\everyone")
-			if Integrations.Discord.CompactMode || overloading {
-				var toSend string
-				if msg.senderName == "" {
-					toSend = strings.ReplaceAll(stripansi.Strip("["+msg.channel+"] "+txt), `\n`, "\n")
-				} else {
-					toSend = strings.ReplaceAll(stripansi.Strip("["+msg.channel+"] **"+msg.senderName+"**: "+txt), `\n`, "\n")
+			if Integrations.Discord.CompactMode {
+				sendDiscordCompactMessage(msg, txt, sess)
+			} else {
+				avatar := createDiscordImage(msg.senderName)
+				avatarHash := shasum(avatar)
+				var webhook *discordgo.Webhook
+
+				// find the webhook for this user
+				for _, wh := range webhooks {
+					if wh.Name == avatarHash {
+						webhook = wh
+						break
+					}
 				}
-				_, err = sess.ChannelMessageSend(Integrations.Discord.ChannelID, toSend)
-				if err != nil {
-					Log.Println("Error sending Discord message:", err)
+				// delete unused webhooks if there are too many and a new one is needed
+				// (discord can only have 15 webhooks per channel)
+				if webhook == nil && len(webhooks) > maxWebhookCount {
+					// generate a list of all users in all rooms
+					users := make([]string, 0, maxWebhookCount)
+					for _, room := range Rooms {
+						for _, user := range room.users {
+							users = append(users, user.Name)
+						}
+					}
+					users = append(users, "", Devbot)
```

Comment:
Doesn't feel right that this list is generated every message. Can we instead just update this list whenever users are added or removed?

Reply:
Also, what happens when plugins send messages?

Reply:
We could have one webhook that handles all miscellaneous situations

Reply:
> Doesn't feel right that this list is generated every message. Can we instead just update this list whenever users are added or removed?

This would need deeper changes outside of the discord bridge... i can try adding this in an expandable style or just for discord 🤷‍♀️

Reply:
> We could have one webhook that handles all miscellaneous situations

i could use the empy name webhook that is used for system commands, or add a difference between the max number of webhooks and the max number of users to use webhooks for so that webhooks are created for plugins, they might just get deleted after a user changes color or another plugin/the same plugin sends a message with different colors...